### PR TITLE
JavaScript: Allow specifying additional remote flow sources through JSON.

### LINKED ISSUE #4785: Interchange format for listing JavaScript sinks
We don't use CodeQL yet at Vimeo, but it could be very useful for detecting XSS vulnerabilities in a TypeScript codebase where developers use the `dangerouslySetInnerHTML` React attribute.

Given a tool that could generate a list of JavaScript sinks like 

```
`window.someObject` in `src/Foo/Bar.ts` contains the unescaped user-controlled fields [`first_name`, `last_name`]
```

Could that list then be imported into CodeQL?

This cross-language taint-tracking would be useful in a few places where SPAs have supplanted legacy HTML (which our existing taint-analysis tool is now able to monitor).

### PR DESCRIPTION
**Pull request by max-schaefer**
Fixes #4785.

With this PR, you can add one or more JSON files named `codeql-javascript-remote-flow-sources.json` to your database (either by including them somewhere in the source tree or by extracting and importing them separately) to specify additional remote flow sources, which will give rise to new `RemoteFlowSource`s at the QL level.

As an example of the syntax, this JSON file

```json
{
    "user input": [
        "window.user.name",
        "window.user.address",
        "window.dob"
    ],
    "uncontrolled path": [
        "window.upload"
    ]
}
```

marks `window.user.name`, `window.user.address`, `window.dob`, and `window.upload` as remote flow sources, the former three with source type "user input", the last one with source type "uncontrolled path".

We use inter-procedural data flow to identify accesses to the new remote flow sources, so a property access of the form `getWindow().upload` where `getWindow()` returns a reference to the global object is also recognised as a remote flow source.

### CODE REVIEW

**Comment by max-schaefer** [comment]
I haven't evaluated performance or written proper documentation yet, but putting this up anyway to facilitate discussion. (@asgerf)

@muglug, would [this format](https://github.com/github/codeql/pull/4802/files#diff-133427b45b7318ee8b922439b93856e02e7480d90d42e3601ec6d708bc09265b) work for you?

The one thing that's currently missing is restricting the scope of a remote flow source to a single source file. I wasn't sure we want that, since the sources are (at the moment anyway) global variables anyway. Do you have specific use cases where you want different sets of (global) flow sources for different files?

**Comment by muglug** [comment]
That format would work great!

**Comment by muglug** [comment]
Is there also a way to list TypeScript types as sources?

e.g.

```ts
export type TeamSettings = {
  teamName: string;  // RemoteFlowSource: user input
  logoUrl: string;
  accentColor: string;
};
```

These types are typically mapped to JSON requests inside application code.

**Comment by max-schaefer** [comment]
Ah, good question. I think that should be possible, but may need a few more extensions.

**Comment by max-schaefer** [comment]
Performance looks [unproblematic](https://github.com/max-schaefer/dist-compare-reports/blob/master/js/external-remote-flow-sources/report.md) (internal link).

I'd suggest we tackle the extension to TypeScript in a separate PR.

Still need to add documentation in the appropriate places, though.

**Review by asgerf — COMMENTED** [review]
Implementation LGTM. 👍

**Comment by max-schaefer** [comment]
I have added a brief write-up in `docs/codeql/codeql-language-guides`.

@mchammer01, could you take a look? (And let me know if this isn't the right location or the right format.)

**Comment by mchammer01** [comment]
Hi @max-schaefer 👋🏻 - Unfortunately, I don't have time to look at this (heads down with documentation work for GHES 3.0) but someone else in the team will review this for you, hopefully shortly 🙂

**Comment by jf205** [comment]
> I have added a brief write-up in `docs/codeql/codeql-language-guides`.
> 
> @mchammer01, could you take a look? (And let me know if this isn't the right location or the right format.)

Hey @max-schaefer

MC is heads down on GitHub docs so I'll take a look at this for you

**Comment by jf205** [comment]
I've just pushed a commit that adds a short introduction to the article, which is also used in the table of contents on the `CodeQL for JavaScript` landing page. Please feel free to tweak the wording if it's not accurate. I'll review the rest of the text shortly.

**Comment by sj** [comment]
Awesome work team! Let's ship this as a beta for the CodeQL JavaScript/TypeScript analysis and learn from how people use it. We'll then consider whether (and how) we can integrate this into the CodeQL analysis for other languages.

@jf205 / @max-schaefer: could you make sure to include the "beta" label in the documentation please?

Thanks all!

File: docs/codeql/codeql-language-guides/specifying-additional-remote-flow-sources-for-javascript.rst [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+To model a new source of untrusted input, such as a previously unmodelled library API, you can
+define a subclass of ``RemoteFlowSource`` that covers all uses of that API. All standard analyses
+will then automatically pick up this new source of remote flow.
+
+However, this approach requires writing QL code and adding it to the standard library, which is not
+always easy to do. Instead, you can also add a JSON file describing custom sources of untrusted
+input to your code base and have it picked up without needing to modify the standard library. This
+JSON file can be hand-written or generated by another tool, but note that these customizations
+are specific to the code base containing the JSON file. To use them in other code bases, you need
+to copy over the JSON file.
```
Comment (jf205) [inline docs/codeql/codeql-language-guides/specifying-additional-remote-flow-sources-for-javascript.rst:27]:
```suggestion
JSON file can be hand-written or generated by another tool. The custom remote flow sources are only available to the code base containing the JSON file. This means that you need to copy the JSON file into each code base that requires the customizations.
```

File: docs/codeql/codeql-language-guides/specifying-additional-remote-flow-sources-for-javascript.rst [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+Specification format
+--------------------
+
+The JSON file must be called ``codeql-javascript-remote-flow-sources.json`` (case-sensitive) and
```
Comment (jf205) [inline docs/codeql/codeql-language-guides/specifying-additional-remote-flow-sources-for-javascript.rst:32]:
```suggestion
The JSON file must be called ``codeql-javascript-remote-flow-sources.json`` and
```

File: docs/codeql/codeql-language-guides/specifying-additional-remote-flow-sources-for-javascript.rst [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+can be located anywhere in your code base. It should consist of a single JSON object. The property
+names of this object are interpreted as `source types`. The values they map to should be arrays of
+strings. Each string should be of the form ``window.props``, where ``props`` is a sequence of one
+or more property names separated by dots, specifying that any value reachable from the global window
+object by this sequence of property names should be considered as untrusted user input of the
+associated source type.
```
Comment (jf205) [inline docs/codeql/codeql-language-guides/specifying-additional-remote-flow-sources-for-javascript.rst:38]:
```suggestion
names of this object are interpreted as `source types`. The values they map to should be arrays of
strings. Each string should be of the form ``window.props``, where ``props`` is a sequence of one
or more property names separated by dots. This notation specifies that any value reachable from the global window
object by this sequence of property names should be considered as untrusted user input of the
associated source type.
```

File: docs/codeql/codeql-language-guides/specifying-additional-remote-flow-sources-for-javascript.rst [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+Example
+-------
+
+As an example, consider the following specification:
```
Comment (jf205) [inline docs/codeql/codeql-language-guides/specifying-additional-remote-flow-sources-for-javascript.rst:43]:
```suggestion
Consider the following specification:
```

**Review by jf205 — APPROVED** [review]
I've added a beta note and reviewed the rest of the text. Looks pretty much good to go 👍🏻

**Comment by max-schaefer** [comment]
Thanks, @jf205!

We may need to iterate a bit more on

> You can model potential sources of untrusted user input in external files

I think this could be misinterpreted as referring to untrusted input coming from external files, which of course isn't what this is about.

File: docs/codeql/codeql-language-guides/specifying-additional-remote-flow-sources-for-javascript.rst [hunk 5]
```diff
@@ -0,0 +1,51 @@
+.. _specifying-additional-remote-flow-sources-for-javascript:
+
+Specifying additional remote flow sources for JavaScript
+========================================================
+
+You can model potential sources of untrusted user input in external files without making changes to the CodeQL standard library.
```
Comment (jf205) [inline docs/codeql/codeql-language-guides/specifying-additional-remote-flow-sources-for-javascript.rst:6]:
```suggestion
You can model potential sources of untrusted user input in your code without making changes to the CodeQL standard library by specifying extra remote flow sources in an external file.
```

File: docs/codeql/codeql-language-guides/codeql-for-javascript.rst [hunk 6]
```diff
@@ -27,6 +28,8 @@ Experiment and learn how to write effective and efficient queries for CodeQL dat
 
 -  :doc:`Using flow labels for precise data flow analysis <using-flow-labels-for-precise-data-flow-analysis>`: You can associate flow labels with each value tracked by the flow analysis to determine whether the flow contains potential vulnerabilities.
 
+-  :doc:`Specifying remote flow sources for JavaScript <specifying-additional-remote-flow-sources-for-javascript>`: You can model potential sources of untrusted user input in external files without making changes to the CodeQL standard library.
```
Comment (jf205) [inline docs/codeql/codeql-language-guides/codeql-for-javascript.rst:31]:
```suggestion
-  :doc:`Specifying remote flow sources for JavaScript <specifying-additional-remote-flow-sources-for-javascript>`: You can model potential sources of untrusted user input in your code without making changes to the CodeQL standard library by specifying extra remote flow sources in an external file.
```

**Review by jf205 — COMMENTED** [review]
Good point. He's another attempt.

**Comment by max-schaefer** [comment]
Perfect, thanks!

**Review by jf205 — APPROVED** [review]
Docs LGTM!

**Review by asgerf — APPROVED** [review]

**Pull request merged**